starport/services/network/join.go

```diff
@@ -0,0 +1,218 @@
+package network
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	launchtypes "github.com/tendermint/spn/x/launch/types"
+	"github.com/tendermint/starport/starport/pkg/events"
+)
+
+// Join creates the RequestAddValidator message into the SPN
+func (b *Builder) Join(
+	ctx context.Context,
+	launchID uint64,
+	peer,
+	valKeyName,
+	valAddress string,
+	gentx,
+	consPubKey []byte,
+	selfDelegation sdk.Coin,
+) (uint64, bool, error) {
+	// Change the chain address prefix to spn
+	spnValAddress, err := SetSPNPrefix(valAddress)
+	if err != nil {
+		return 0, false, err
+	}
+
+	// Check if the validator request already exist
+	exist, err := b.CheckValidatorExist(ctx, launchID, spnValAddress)
+	if err != nil {
+		return 0, false, err
+	}
+	if exist {
+		return 0, false, fmt.Errorf("validator %s already exist", spnValAddress)
+	}
+
+	msg := launchtypes.NewMsgRequestAddValidator(
+		spnValAddress,
+		launchID,
+		gentx,
+		consPubKey,
+		selfDelegation,
+		peer,
+	)
+
+	b.ev.Send(events.New(events.StatusOngoing, "Broadcasting validator transaction"))
+	res, err := b.cosmos.BroadcastTx(valKeyName, msg)
+	if err != nil {
+		return 0, false, err
+	}
+
+	var requestRes launchtypes.MsgRequestResponse
+	if err := res.Decode(&requestRes); err != nil {
+		return 0, false, err
+	}
+	b.ev.Send(events.New(events.StatusDone, "AddValidator transaction sent"))
+
+	return requestRes.RequestID, requestRes.AutoApproved, nil
+}
+
+// CreateAccountRequestMsg creates an add AddAccount request message
+func (b *Builder) CreateAccountRequestMsg(
+	ctx context.Context,
+	chainHome string,
+	customGentx bool,
+	launchID uint64,
+	amount sdk.Coin,
+) (uint64, bool, bool, error) {
+	address := b.account.Address(SPNAddressPrefix)
+	spnAddress, err := SetSPNPrefix(address)
+	if err != nil {
+		return 0, false, false, err
+	}
+
+	b.ev.Send(events.New(events.StatusOngoing, "Verifying account already exists "+spnAddress))
+
+	shouldCreateAcc := false
+	if !customGentx {
+		exist, err := CheckGenesisAddress(chainHome, spnAddress)
+		if err != nil {
+			return 0, false, false, err
+		}
+		if !exist {
+			exist, err = b.CheckAccountExist(ctx, launchID, spnAddress)
+			if err != nil {
+				return 0, false, false, err
+			}
+		}
+		shouldCreateAcc = !exist
+	}
+	if shouldCreateAcc || customGentx {
```

Comment:
why `customGentx` means an account must be created?

Reply:
I put the verification in the wrong place, its fixed now